extensions/amp-story-auto-ads/0.1/story-ad-page.js

```diff
@@ -0,0 +1,486 @@
+
+    /** @private {?Document} */
+    this.adDoc_ = null;
+
+    /** @private {?string} */
+    this.ctaText_ = null;
+
+    /** @private {?string} */
+    this.ctaUrl_ = null;
+
+    /** @private {?string} */
+    this.ampAdExitOutlink_ = null;
+
+    /** @private {boolean} */
+    this.loaded_ = false;
+
+    /** @private @const {!JsonObject} */
+    this.a4aVars_ = dict();
+
+    /** @private @const {!Array<Function>} */
+    this.loadCallbacks_ = [];
+  }
+
+  /** @return {?Document} ad document within FIE */
+  getAdDoc() {
+    return this.adDoc_;
+  }
+
+  /** @return {string} */
+  getId() {
+    return this.id_;
+  }
+
+  /** @return {boolean} */
+  hasTimedOut() {
+    return (
+      !!this.timeCreated_ && Date.now() - this.timeCreated_ > TIMEOUT_LIMIT
+    );
+  }
+
+  /** @return {boolean} */
+  isLoaded() {
+    return this.loaded_;
+  }
+
+  /** @return {?Element} */
+  getPageElement() {
+    return this.pageElement_;
+  }
+
+  /**
+   * Register functions to be executed when ad has loaded.
+   * @param {Function} cb
+   */
+  registerLoadCallback(cb) {
+    this.loadCallbacks_.push(cb);
+  }
+
+  /**
+   * Adds/removes [amp-story-visible] on FIE body so that animations can
+   * respond accordingly.
+   */
+  toggleVisibility() {
+    // TODO(calebcordry): Properly handle visible attribute for custom ads.
+    if (this.adDoc_) {
+      toggleAttribute(
+        dev().assertElement(this.adDoc_.body),
+        PageAttributes.IFRAME_BODY_VISIBLE
+      );
+      // TODO(#24829) Remove alternate body when we have full ad network support.
+      const alternateBody = this.adDoc_.querySelector('#x-a4a-former-body');
+      alternateBody &&
+        toggleAttribute(alternateBody, PageAttributes.IFRAME_BODY_VISIBLE);
+    }
+  }
+
+  /**
+   * Create an `amp-story-page` containing an `amp-ad`.
+   * @return {!Element}
+   */
+  build() {
+    this.timeCreated_ = Date.now();
+    this.pageElement_ = this.createPageElement_();
+    this.adElement_ = this.createAdElement_();
+
+    const glassPane = this.doc_.createElement('div');
+    glassPane.classList.add(GLASS_PANE_CLASS);
+
+    const gridLayer = this.doc_.createElement('amp-story-grid-layer');
+    gridLayer.setAttribute('template', 'fill');
+
+    const paneGridLayer = gridLayer.cloneNode(/* deep */ false);
+
+    gridLayer.appendChild(this.adElement_);
+    paneGridLayer.appendChild(glassPane);
+    this.pageElement_.appendChild(gridLayer);
+    this.pageElement_.appendChild(paneGridLayer);
+
+    // Set up listener for ad-loaded event.
+    this.adElement_
+      .signals()
+      // TODO(ccordry): Investigate using a better signal waiting for video loads.
+      .whenSignal(CommonSignals.INI_LOAD)
+      .then(() => this.whenAdLoaded_());
+
+    this.analyticsEvent_(AnalyticsEvents.AD_REQUESTED, {
+      [AnalyticsVars.AD_REQUESTED]: Date.now(),
+    });
+
+    return this.pageElement_;
+  }
+
+  /**
+   * @return {!Element}
+   * @private
+   */
+  createPageElement_() {
```

Comment:
code style: move private methods to the bottom

Reply:
moved `maybeCreateCTA` up. Looks like all the other public methods were already at top.